### PR TITLE
feat: Added grouping in default_config page

### PR DESCRIPTION
**Pull request by mahatoankitkumar**
## Problem
Add grouping in default_config page

## Solution
Added grouping in default_config page to show keys in folder structure.
Screenshot 2024-04-23 at 5 35 57 PM
Screenshot 2024-04-23 at 5 36 04 PM
Screenshot 2024-04-23 at 5 36 14 PM

## Environment variable changes
N.A.

## Pre-deployment activity
N.A.

## Post-deployment activity
N.A.

## API changes
N.A.

## Possible Issues in the future
N.A.

### CODE REVIEW

**Comment by Datron** [comment]
@mahatoankitkumar can you add screenshots?

File: crates/frontend/src/pages/DefaultConfig/DefaultConfig.rs [hunk 1]
```diff
@@ -184,23 +242,117 @@ pub fn DefaultConfig() -> impl IntoView {
...
-                                <div class="flex justify-between">
-                                    <h2 class="card-title chat-bubble text-gray-800 dark:text-white bg-white font-mono">
-                                        "Default Config"
-                                    </h2>
-                                    <DrawerBtn drawer_id="default_config_drawer".to_string()>
-                                        Create Key <i class="ri-edit-2-line ml-2"></i>
-                                    </DrawerBtn>
+                                <div class="flex justify-between pb-2">
+                                    <div class="flex justify-between pt-3">{
+                                        {move || {
```
Comment (ShubhranshuSanjeev) [inline crates/frontend/src/pages/DefaultConfig/DefaultConfig.rs:296]:
```suggestion
```suggestion
                                        {
                                            move || {
                                                let data = bread_crums.get();
                                                let last_index = data.len() - 1;
                                                data.iter()
                                                    .enumerate()
                                                    .map(|(index, item)| {
                                                        let value = item.value.clone();
                                                        let is_link = item.is_link.clone();
                                                        let is_last_item = index < last_index;
                                                        let item_classname = if item.is_link {
                                                            "cursor-pointer text-blue-500 underline underline-offset-2"
                                                        } else {
                                                            ""
                                                        };
                                                        view! {
                                                            <div class="flex">
                                                                <h2
                                                                    on:click=move |_| {
                                                                        if is_link {
                                                                            folder_click_handler(value.clone())
                                                                        }
                                                                    }
                                                                    class={item_classname}
                                                                >
                                                                    {item.key.clone()}
                                                                </h2>
                                                                <h2 class="pl-4 pr-4">
                                                                    {if is_last_item { ">" } else { "" }}
                                                                </h2>
                                                            </div>
                                                        }
                                                    })
                                                    .collect_view()
                                            }
                                        }


```
```

Comment (ShubhranshuSanjeev) [reply crates/frontend/src/pages/DefaultConfig/DefaultConfig.rs:296]: We can use derive signal like this, to solution this out

```
    let bread_crums = Signal::derive(move || {
        let filters = filters.get();
        let mut default_bread_crums = vec![BreadCrums {
            key: "Default Config".to_string(),
            value: None,
            is_link: true,
        }];

        let mut bread_crums = match filters {
            Some(prefix) => {
                let partials = prefix
                    .trim_matches('.')
                    .split(".")
                    .map(str::to_string)
                    .collect::<Vec<String>>();
                partials
                    .into_iter()
                    .fold(String::new(), |mut prefix, partial| {
                        default_bread_crums.push(BreadCrums {
                            key: partial.clone(),
                            value: Some(prefix.clone()),
                            is_link: true,
                        });
                        prefix.push_str(&partial);
                        prefix.push('.');
                        prefix
                    });
                default_bread_crums
            }
            None => default_bread_crums,
        };

        if let Some(last_crum) = bread_crums.last_mut() {
            last_crum.is_link = false;
        }

        bread_crums
    });

```

**Review by ShubhranshuSanjeev — CHANGES_REQUESTED** [review]

File: crates/frontend/src/components/default_config_form/default_config_form.rs [hunk 2]
```diff
@@ -23,6 +23,7 @@ pub fn default_config_form<NF>(
     #[prop(default = String::new())] config_pattern: String,
     #[prop(default = String::new())] config_value: String,
     #[prop(default = None)] function_name: Option<Value>,
+    #[prop(default = None)] prefix: Option<String>,
```
[the remark this answers is not in the document]
Comment (mahatoankitkumar) [reply crates/frontend/src/components/default_config_form/default_config_form.rs:26]: Optional parameter makes more sense here.

File: crates/frontend/src/components/default_config_form/default_config_form.rs [hunk 3]
```diff
@@ -65,7 +66,9 @@ where
 
     let on_submit = move |ev: MouseEvent| {
         ev.prevent_default();
-        let f_name = config_key.get();
+        let f_name = prefix
+            .clone()
+            .map_or_else(|| config_key.get(), |p| p + &config_key.get());
```
[the remark this answers is not in the document]
Comment (mahatoankitkumar) [reply crates/frontend/src/components/default_config_form/default_config_form.rs:71]: Adds prefix to the key. Will change `p` to `prefix` more readibility.

File: crates/frontend/src/pages/default_config/mod.rs [hunk 4]
```diff
@@ -0,0 +1 @@
+pub mod default_config;
```
Comment (Datron) [inline crates/frontend/src/pages/default_config/mod.rs:1]: Can we stop doing these one line mod.rs files? We can just put all the code here

Comment (Datron) [reply crates/frontend/src/pages/default_config/mod.rs:1]: Leave it as is for now

File: crates/experimentation_platform/src/db/models.rs [hunk 5]
```diff
@@ -1,12 +1,21 @@
...
 #[derive(
-    Debug, Clone, Copy, PartialEq, Deserialize, Serialize, diesel_derive_enum::DbEnum, QueryId,
+    Debug,
+    Clone,
+    Copy,
+    PartialEq,
+    Deserialize,
+    Serialize,
+    diesel_derive_enum::DbEnum,
+    QueryId,
```
Comment (Datron) [inline crates/experimentation_platform/src/db/models.rs:18]: @mahatoankitkumar you can get rid of QueryId here, I was trying to solve something and added it by mistake.

**Review by Datron — APPROVED** [review]

File: crates/frontend/src/pages/default_config/default_config.rs [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+    let query_params = use_query_map();
+    let bread_crums = Signal::derive(move || {
+        let key_prefix = key_prefix.get();
+        let mut default_bread_crums = vec![BreadCrums {
+            key: "Default Config".to_string(),
+            value: None,
+            is_link: true,
+        }];
+
+        let mut bread_crums = match key_prefix {
```
Comment (pratikmishra356) [inline crates/frontend/src/pages/default_config/default_config.rs:46]: can we have different variable name `bread_crums` ?

Comment (mahatoankitkumar) [reply crates/frontend/src/pages/default_config/default_config.rs:46]: bread_crums name serves the purpose.

**Review by ShubhranshuSanjeev — CHANGES_REQUESTED** [review]

File: crates/frontend/src/pages/default_config/default_config.rs [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+                                <Stat heading="Config Keys" icon="ri-tools-line" number=total_default_config_keys/>
+                            </div>
+                            <div class="card rounded-lg w-full bg-base-100 shadow">
+                                <div class="card-body">
+                                    <div class="flex justify-between pb-2">
+                                        <BreadCrums bread_crums=bread_crums.get()  folder_click_handler/>
+                                        <div class="flex">
+                                            <label on:click=move |_| {
+                                                folder_click_handler(None);
+                                                enable_grouping.set(!enable_grouping.get());}
```
[the remark this answers is not in the document]
Comment (mahatoankitkumar) [reply crates/frontend/src/pages/default_config/default_config.rs:269]: This isn't re-rendering the component.

**Review by ShubhranshuSanjeev — APPROVED** [review]

**Pull request merged**